code/game/gamemodes/umbrage/umbrage_datum.dm

```diff
@@ -0,0 +1,136 @@
+//Handles everything related to psi and lucidity.
+/datum/umbrage
+	var/datum/mind/linked_mind
+	var/mob/living/linked_body
+	var/psi = 100 //Our psi, used for abilities.
+	var/max_psi = 100
+	var/psi_regeneration = 20 //The limit for psi regenerated during the cycle.
+	var/psi_used_since_last_cycle //How much psi we've used in the last five seconds.
+	var/cycle_progress = 0 //When this reaches 5, it will reset to 0 and regenerate up to 20 spent psi.
+	var/lucidity = 3 //Lucidiy is used to buy abilities. We gain one each time we drain someone.
+	var/lucidity_drained = 3 //How much lucidity we've taken overall.
+	var/list/linked_abilities = list() //Abilities that we have
+	var/list/linked_ability_types = list() //Used for cloning and gaining new bodies
+	var/list/drained_minds = list() //Minds eaten with Devour Will.
```

Comment:
Ok so, byond optimises memory very well (MSO once said something along the lines that if byond weren't as good with memory as it is, the game would take 4x the memory to run), but it sacrifices CPU for this, and most of the time we prefer having good CPU usage, which is why we come up with cool shit like typecaches, which are a memory-hungry, cpu-friendly data structure (or atleast, usage of an existing structure, assoc lists)

The less memory we waste, the more we can cautiously spend to help out with CPU, and unwasted memory is just free memory to use in other ways too.